### ISSUE TITLE
Undo restores incorrect ctx

### ISSUE THREAD

**Issue by mgrau** [nicolodavis/boardgame.io#389]
When performing an undo operation, ctx takes the value of 2 moves prior, instead of 1 move prior. This is particularly problematic when a move changes the phase. I've found that using
```
        return {
          ...state,
          G: restore.G,
          ctx: last.ctx,
          _undo: _undo.slice(0, _undo.length - 1),
          _redo: [last, ..._redo],
        };
```
seems to give the correct behavior.

**Comment by nicolodavis** [comment]
Hey @mgrau, I thought we tested this feature extensively when we first built it: https://github.com/nicolodavis/boardgame.io/blob/master/src/core/reducer.test.js#L267

1. Do the unit tests I've linked to match your expectations?
2. Do they still match your expectations after changing the code to your proposal?

I'm not in front of a computer at the moment so can't try this out myself yet.

**Comment by nicolodavis** [comment]
But then `G` and `ctx` would come from different snapshots, which doesn't make sense.  The last element of `_undo` (i.e. `_undo[length-1]`) is the current game state, so it makes sense to restore to `_undo[length - 2]`.

I tried to reproduce your problem with some examples, but I am not able to (things seem to be working correctly, even when the phase changes).

Also note that we have a bunch of unit tests for the undo logic [here](https://github.com/nicolodavis/boardgame.io/blob/master/src/core/reducer.test.js#L267).  Can you demonstrate your problem via an additional test here?

**Comment by mgrau** [comment]
I agree, it doesn't make sense to use G and ctx from different game states.
I need to look around more to see where else ctx is changing.

I noticed that if I restore ctx to _undo[_undo.length - 1].ctx, the undo
unit tests still pass. I'll focus on trying to replicate my problem with a
test.

**Comment by nicolodavis** [comment]
I think that's probably because the tests only verify the state in `G`.  Maybe you can add a test that also checks `ctx` to see if all looks good (I'm happy to get that in regardless of whether a bug exists).

**Comment by mgrau** [comment]
It looks like `G` and `ctx` are appended to `_undo` in `processMove()`, however, this is done before `ctx` is updated by `updateAndDetach()`, which means that the current `ctx` is not reflected in `_undo`. I don't think this is intentional. I've made a test that highlights this behavior. 

I'm not sure the best place to fix this, but I've added 
```
state._undo[state._undo.length - 1].ctx = state.ctx;
```
after ctx is updated by `updateAndDetach()`.
